### ISSUE TITLE
Live hls when paused don't fill up the buffer to DEFAULT_MAX_BUFFER_MS

### ISSUE THREAD

**Issue by fgl27** [google/ExoPlayer#7006]
### [REQUIRED] Issue description
Hls live streams that I leave it paused for a time that is bigger than **playlist.durationUs** don't fill up DEFAULT_MAX_BUFFER_MS.

It doesn't loading new segments after the buffer size became current buffer (before pause buffer) + **playlist.durationUs** or something like that in my case the before buffet is 5 to 10s and the playlist.duration is 28 to 32s, the buffer stops incrementing at around 35 to 45s of buffer size, none new segments are buffered because the playlist is not updated after that during a pause on live streams, even if **DEFAULT_MAX_BUFFER_MS** is "not yet full".

The result is, that after I press play it plays for **buffer size**, until the buffer is empty, then it creates a discontinuity because it "jumps back to the live feed" (after a new buffer moment) without ever trying to get the missing segment's...

I would like to see **DEFAULT_MAX_BUFFER_MS** to get full on pause for all live stream, even if **playlist.durationUs** and the visible segments from before I click pause are less then **DEFAULT_MAX_BUFFER_MS**, after all for a live stream segments keep be generated even if I pause it...

### [REQUIRED] Reproduction steps
Open a live stream that has a limited value for **playlist.durationUs**, pause the stream the buffer will not fill up to DEFAULT_MAX_BUFFER_MS

### [REQUIRED] Link to test content
Any live hls stream with a limited list duration in seconds will do,the ones I have expire very fast.

### [REQUIRED] A full bug report captured from the device
Not applicable as noting is crashing.

### [REQUIRED] Version of ExoPlayer being used
2.11.3

### [REQUIRED] Device(s) and version(s) of Android being used
All devices I tested show this problem, Android 8 and up.

**Comment by phhusson** [comment]
The Loading Control algorithm can be viewed here:
https://github.com/google/ExoPlayer/blob/90b62c67fbf29bff2406d16979539da4ec126166/library/core/src/main/java/com/google/android/exoplayer2/DefaultLoadControl.java#L384-L400

Personnally for live streaming I wrote my own LoadControl, precisely for this use-case, because you always want to continue loading, and you probably want specific behaviour about backbuffer.


Anyway, assuming you keep using DefaultLoadControl, you can see with the function I linked why this is doing that:
The max is really just a max that might be never reached. The actual buffer size it is trying to target is computed using calculateTargetBufferSize, or you can overload it with setTargetBufferBytes.
What you probably want is to set setTargetBufferBytes to the amount of RAM you want to dedicate to timeshifting, and set maxBufferMs to "infinity".

**Comment by tonihei** [comment]
We already have a pending change to update the target buffer size to match our time constraints (for most playbacks). This should fix the mentioned issue. You can also try to set the prioritizeTimeOverSizeThresholds config value to false for a quick workaround.

**Comment by fgl27** [comment]
Yes I was using the DefaultLoadControl wrong... With the help of you guys I can see that I can load more seconds they before.

I use the technique of overwriting the **setTargetBufferBytes** and works OK, I was using it but wrong before now I understood it.

> What you probably want is to set setTargetBufferBytes to the amount of RAM you want to dedicate to timeshifting, and set maxBufferMs to "infinity".

infinity? what is the real value accepted here?

I know that there is a "max" possible for a int, but that value is the maximum I can use here or exo has a maximum value define somewhere?

**Comment by fmatosqg** [comment]
TL;DR I'm ok that pause is the default behaviour but I need fallback behaviour to keep playing forever no matter what error happens (well, almost all errors) - so I'd like something like an opt-in instead of having to write a lot of (hard to test) code to understand and deal with every error scenario - at least for `BehindLiveWindowException`. 


The primary purpose of my app is live streaming from a given url. We're a traditional radio broadcast company and our HLS window is about 30 seconds, because it's synched to the live stream over FM / AM.

Realistically users tend to use the streaming feature on average 30 / 40 minutes per session. My goal / requirement is to be able to stream over 10h no matter network handover / battery / doze mode conditions on even the most strict phone brand. With this I'd hope we get less complaints about audio stopping unexpectedly from real users.

Math and windows apart, these are some of the situations I noticed that can trigger a `BehindLiveWindowException`:

1. network connectivity is glitchy, and audio pauses automatically until network is back (I also have code to detect and recover from it). Obviously it's quite common that sometimes windows won't be available anymore, so I'd like to keep playing anything rather than stop / pause.
1. nothing wrong is noticed on network (or so briefly that you can't hear audio pauses or clicks) but after a long session I get `BehindLiveWindowException`.
1.  user manually pauses and unpauses streaming. Current window finishes playing and then I get `BehindLiveWindowException`. Again, stopping / pausing automatically or by error detection is not an option, and it's preferrable to play anything as long as the user is not bothered to intervene. So my code recovers from that by starting to play again from same url.

I wouldn't expect my users to understand HLS windows and the effect of pausing when they unpause. And as a side note, I have also noticed case 1 when the network seems to be fine ( I have added logs and calls to `TextToSpeech` to many error conditions to be able to diagnose things on the fly).

I hope these notes apply to this issue, but happy to move it somewhere else if not the case.

**Comment by fgl27** [comment]
@fmatosqg on the past I was getting that exception, was because the parameter `#EXT-X-TARGETDURATION` the value of it in the list was and is today bigger than individual segments size.

That was discussed here https://github.com/google/ExoPlayer/issues/6360 some improve was made but in the end I just have to overwrite target duration and for me is easier to building exo from source as I do other changes so I do this https://github.com/fgl27/ExoPlayer/blob/dev-v2/library/hls/src/main/java/com/google/android/exoplayer2/source/hls/playlist/HlsMediaPlaylist.java#L274

As long the value of `targetDurationUs` is lower than a individual segments my playback is fine. 

So check that also maybe you have a similar issue on some of those hls list you have...

**Comment by fmatosqg** [comment]
@fgl27 My problem seems to be very different from yours, and I wouldn't do a custom build for exoplayer since my BehindLiveException problem is solved - though at a cost, and I have a lake full of bigger fish to fry, like doze vs foreground services.

My point is that it should be very easy for a dev to handle  BehindLiveException and many other errors. Something as simple as `player.play(handleBehindLiveException = true)`, or in audio attributes or similar. No more than 1 line of code, no listener, no callbacks ;)

**Comment by tonihei** [comment]
I'm not sure whether this issue is still tracking anything useful.

 - `DefaultLoadControl` sets the maximum allowed buffer size in bytes so that almost all streams should be able to fill the buffer up to the maximum buffer duration in ms. The default of `prioritizeTimeOverSizeThresholds` is also set to false (see commit above).
 - The buffer durations, max buffer size in bytes and which value to prioritize can all be configured depending on the app needs.
 - HLS playbacks fill up their buffer up the configured max buffer as long as the next segment to load is in the currently available live window. 

Closing this issue under the assumption that everything is either working or can be configured. If there is still something to be done here, please let us know.
